src/realm/query_expression.hpp

```diff
@@ -487,80 +488,156 @@ class ValueBase {
+
+        if (left.m_from_list && right.m_from_list && !left_cmp_type && !right_cmp_type) {
+            // Both lists and no ANY, NONE, ALL specified - simple element by element comparison
+            if (left.size() != right.size()) {
+                if constexpr (std::is_same_v<TCond, NotEqual>) {
+                    return 0; // mismatch size
+                }
+                else {
+                    return not_found;
+                }
+            }
+            for (size_t i = 0; i < left.size(); ++i) {
+                if (!c(left[i], right[i])) {
+                    return not_found;
+                }
+            }
+            return 0; // all elements matched in the right order
         }
-        else if (!left.m_from_link_list && right.m_from_link_list) {
-            // Right values come from link list. Left must come from single row. Semantics: Match if at least 1
-            // linked-to-value fulfills the condition
-            REALM_ASSERT_DEBUG(left.size() > 0);
-            const size_t num_right_values = right.size();
-            ValueType left_val = left[0];
-            for (size_t r = 0; r < num_right_values; r++) {
-                bool match = c(left_val, right[r]);
-                if (match) {
-                    if (right_cmp_type == ExpressionComparisonType::Any) {
+
+        // if one side omitted a comparison type, assume ANY
+        const Compare compare_left = left_cmp_type.value_or(Compare::Any);
+        const Compare compare_right = right_cmp_type.value_or(Compare::Any);
+
+        size_t left_size = left.m_from_list ? left.size() : 1;
+        size_t right_size = right.m_from_list ? right.size() : 1;
+        // remove duplicates to reduce comparison time in nested loops
+        if (left_size > 2 && right_size > 2) {
+            if constexpr (realm::is_any_v<TCond, Greater, GreaterEqual>) {
+                std::sort(left.begin(), left.end());
+                std::sort(right.begin(), right.end(), std::greater<QueryValue>());
+            }
+            else if constexpr (realm::is_any_v<TCond, Less, LessEqual>) {
+                std::sort(left.begin(), left.end(), std::greater<QueryValue>());
+                std::sort(right.begin(), right.end());
+            }
+            else {
+                std::sort(left.begin(), left.end());
+                std::sort(right.begin(), right.end());
+            }
+            left_size = std::unique(left.begin(), left.end()) - left.begin();
+            right_size = std::unique(right.begin(), right.end()) - right.begin();
+        }
+
+        if constexpr (realm::is_any_v<TCond, BeginsWith, BeginsWithIns, EndsWith, EndsWithIns, Contains, ContainsIns,
+                                      Like, LikeIns>) {
+            // The string operators have the arguments reversed so we have to iterate right in the
+            // outer loop as this is actually the left argument
+            auto left_matches = [&](const QueryValue& right_val) {
```

Comment:
It looks like left_matches and right_matches are the exact same function but taking different parameters? Unless there is actually a difference they should just be a function.

Reply:
They are not the same. 'left_matches' uses the parameter as the right hand side in the comparison function whereas 'right_matches' uses the parameter as the right hand side. I am sure there is a way we could avoid duplicating the 'if' control logic, but it could turn out to be a bit convoluted.